src/goto-instrument/contracts/utils.cpp

```diff
@@ -163,3 +167,67 @@ void disable_pointer_checks(source_locationt &source_location)
   source_location.add_pragma("disable:pointer-primitive-check");
   source_location.add_pragma("disable:pointer-overflow-check");
 }
+
+void simplify_gotos(goto_programt &goto_program, namespacet &ns)
+{
+  for(auto &instruction : goto_program.instructions)
+  {
+    if(instruction.is_goto())
+    {
+      const auto &condition = instruction.get_condition();
+      const auto &simplified = simplify_expr(condition, ns);
+      if(simplified.is_false())
+        instruction.turn_into_skip();
+    }
+  }
+}
+
+bool is_loop_free(
+  const goto_programt &goto_program,
+  namespacet &ns,
+  messaget &log)
+{
+  // create cfg from instruction list
+  cfg_baset<empty_cfg_nodet> cfg;
+  cfg(goto_program);
```

Comment:
Comment from the earlier PR: do you really need such an elaborate analysis, or would looking at `is_backwards_goto()` for all instructions in `goto_program` also do the trick?

Reply:
I ignored the existence of `is_backwards_goto`, if it is just jumping to an earlier instruction in the sequence it would not necessarily introduce a loop, would it ?

Reply:
Hmm, that indeed does not imply the existence of a loop. But then, would `natural_loopst` do the trick here and relieve you of some of the heavy lifting?